### DISCUSSION TITLE
Exposing hydration utilities without dependency on React

### DISCUSSION THREAD

**Discussion by DamianOsipiuk** [TanStack/query#2370] [Ideas]
Currently, hydration utilities are exposed together with react hooks. This creates a dependency on React which is unwanted for different frameworks.

Ideally, hydration utilities would be exposed separately. Either by moving them to the core, or exposing them as a separate entry with a rollup.

https://github.com/tannerlinsley/react-query/blob/master/src/hydration/index.ts

Does it make sense to move `hydrate` and `dehydrate` to the core?

Currently, a workaround is to keep an exact copy of the utility implementation.

**Comment by DamianOsipiuk** [comment]
@tannerlinsley @TkDodo

**Comment by TkDodo** [comment]
Is it not it’s separate rollup entry?

https://github.com/tannerlinsley/react-query/blob/master/rollup.config.js#L22

**Reply by DamianOsipiuk** [reply]
Unfortunately react hooks are also exported from index file, which enforces react as a dependency https://github.com/tannerlinsley/react-query/blob/master/src/hydration/react.tsx

**Reply by TkDodo** [reply]
I see. In that case, we'd need to move it to the core I think. I don't know enough about this though, @tannerlinsley or @Ephem maybe?

**Reply by Ephem** [reply]
The hydration tools were written back when RQ only supported React and they were a bit experimental at the start. This was discussed at length back then and they were moved back and forth between core and a separate entry point for exactly this reason.

I argued for keeping it in core since `hydrate` and `dehydrate` are tiny by themselves, but we eventually landed on keeping it in a separate entrypoint, at least initially, to evaluate the API which made sense at the time.

Now that the API is a stable part of RQ and RQ aims to support more frameworks I think it makes total sense to move those utilities into core!

**Reply by Ephem** [reply]
To follow up on this, while the base utilities are small, the framework specific integration might not be depending on framework, so there is also the question of if these should be separate entrypoints or live directly in the framework entrypoints.

For this reason, another approach could of course be to keep hydration entirely separate, but with one entry per framework. Core utils could live in a internal shared package that each framework specific entrypoint imports.

Moving the utils into core or adding separate entrypoints per framework could probably happen in a minor btw as long as we re-export them in `react-query/hydration` and keep that around at least until next major.

**Reply by TkDodo** [reply]
> Moving the utils into core or adding separate entrypoints per framework could probably happen in a minor btw as long as we re-export them in react-query/hydration and keep that around at least until next major.

sounds good, @DamianOsipiuk would you like to give that a try?

**Reply by DamianOsipiuk** [reply]
Yup, will do :+1:
